core/resource/resource.go

```diff
@@ -0,0 +1,255 @@
+		spec:         spec,
+		metadata:     meta,
+		status:       StatusUnknown,
+		existInStore: false,
+	}, nil
+}
+
+func (r *Resource) Name() Name {
+	return r.name
+}
+
+func (r *Resource) FullName() string {
+	if r.kind == KindDataset {
+		return r.dataset.FullName()
+	}
+	return r.dataset.FullName() + "." + r.name.String()
+}
+
+func (r *Resource) URN() string {
+	if r.kind == KindDataset {
+		return r.dataset.URN()
+	}
+	return r.dataset.URN() + "." + r.name.String()
+}
+
+func (r *Resource) Metadata() *Metadata {
+	return r.metadata
+}
+
+func (r *Resource) Kind() Kind {
+	return r.kind
+}
+
+func (r *Resource) Tenant() tenant.Tenant {
+	return r.tenant
+}
+
+func (r *Resource) Dataset() Dataset {
+	return r.dataset
+}
+
+func (r *Resource) Status() Status {
+	return r.status
+}
+
+func (r *Resource) Spec() map[string]any {
+	return r.spec
+}
+
+func (r *Resource) ExistInStore() bool {
+	return r.existInStore
+}
+
+func (r *Resource) Validate() error {
+	switch r.kind {
+	case KindTable:
+		table, err := ConvertSpecTo[Table](r)
+		if err != nil {
+			return err
+		}
+		table.Name = r.name
+		table.Dataset = r.dataset
+		return table.Validate()
+
+	case KindExternalTable:
+		externalTable, err := ConvertSpecTo[ExternalTable](r)
+		if err != nil {
+			return err
+		}
+		externalTable.Name = r.name
+		externalTable.Dataset = r.dataset
+		return externalTable.Validate()
+
+	case KindView:
+		view, err := ConvertSpecTo[View](r)
+		if err != nil {
+			return err
+		}
+		view.Name = r.name
+		view.Dataset = r.dataset
+		return view.Validate()
+
+	case KindDataset:
+		ds, err := ConvertSpecTo[DatasetDetails](r)
+		if err != nil {
+			return err
+		}
+		ds.Dataset = r.dataset
+		return ds.Validate()
+
+	default:
+		return errors.InvalidArgument(EntityResource, "unknown kind")
+	}
+}
+
+func (r *Resource) Equal(incoming *Resource) bool {
+	if r == nil || incoming == nil {
+		return r == nil && incoming == nil
+	}
+	if r.name != incoming.name {
+		return false
+	}
+	if r.kind != incoming.kind {
+		return false
+	}
+	if r.dataset != incoming.dataset {
+		return false
+	}
+	if !reflect.DeepEqual(r.tenant, incoming.tenant) {
+		return false
+	}
+	if !reflect.DeepEqual(r.spec, incoming.spec) {
+		return false
+	}
+	return reflect.DeepEqual(r.metadata, incoming.metadata)
+}
+
+func (r *Resource) ChangeStatusTo(newStatus Status) {
+	r.status = newStatus
+}
+
+func (r *Resource) MarkExistInStore() {
+	r.existInStore = true
+}
+
+type FromExistingOpt func(r *Resource)
+
+func ReplaceStatus(status Status) FromExistingOpt {
+	return func(r *Resource) {
+		r.status = status
+	}
+}
+
+func FromExisting(existing *Resource, opts ...FromExistingOpt) *Resource {
+	output := &Resource{
+		name:         existing.name,
+		kind:         existing.kind,
+		dataset:      existing.dataset,
+		tenant:       existing.tenant,
+		spec:         existing.spec,
+		metadata:     existing.metadata,
+		status:       existing.status,
+		existInStore: existing.existInStore,
```

Comment:
Why do we need this condition ?

Reply:
during the request operation, at some points we will need to confirm whether the particular resource actually exist in datastore or not. take an example when we want to create a table in bigquery. during the create request, we will need to do this check on the bigquery, and accept or reject the request based on it. we need to do so because sometime the resource could already exist, be it from manual creation or "other" optimus or whatnot. this `existInStore` is to help reducing the request to do the check in the actual bigquery.